cpp/src/arrow/engine/substrait/serde_test.cc

```diff
@@ -439,12 +444,36 @@ TEST(Substrait, FieldRefRoundTrip) {
   }
 }
 
+TEST(Substrait, RecursiveFieldRef) {
+  FieldRef ref("struct", "str");
+
+  ARROW_SCOPED_TRACE(ref.ToString());
+  ASSERT_OK_AND_ASSIGN(auto expr, compute::field_ref(ref).Bind(*kBoringSchema));
+  ASSERT_OK_AND_ASSIGN(auto serialized, SerializeExpression(expr));
+  auto expected = SubstraitFromJSON("Expression", R"({
+    "selection": {
+      "directReference": {
+        "structField": {
+          "field": 12,
+          "child": {
+            "structField": {
+              "field": 1
+            }
+          }
+        }
+      },
+      "rootReference": {}
+    }
+  })");
```

Comment:
Please add a test which exercises references into an expression. Something like the following:

```c++
  ASSERT_OK_AND_ASSIGN(expr, compute::call("struct_field", {
    compute::call("if_else", {
      literal(true),
      field_ref("struct"),
      field_ref("struct"),
    })},
    compute::StructFieldOptions{0}).Bind(*kBoringSchema));

  ASSERT_OK_AND_ASSIGN(serialized, SerializeExpression(expr));
  auto expected = SubstraitFromJSON("Expression", R"({
    "selection": {
      "directReference": {
        // get field 0
        "structField": {"field": 0}
      },
      // ... of the following expression
      "expression": {
        "if_then": {
          "ifs": [
            // ... which will always evaluate to the value of field 12
            {
              "if": {"literal": {"boolean": true}},
              "then": {"selection": {"directReference": {"structField": {"field": 12}}}}
            }
          ],
          "else": {"selection": {"directReference": {"structField": {"field": 12}}}}
        }
      }
    }
  })");
```

Reply:
... hardcoding the index 12 is a bit fragile, but we can leave that for follow up.

Reply:
747cbbc